### DISCUSSION TITLE
Grab bag of questions

### DISCUSSION THREAD

**Discussion by governmentSponsored** [uswds/uswds#5523] [General]
## Guidance
1. What is the guidance for styling columns of text data, without putting them into tables?
2. Is there guidance for dynamically displaying alerts based on updates to the site?
3. Do you have accessibility guidelines for tab order? We found [some](https://www.section508.gov/content/guide-accessible-web-design-development/#212-no-keyboard-trap) [things](https://www.section508.gov/content/guide-accessible-web-design-development/#331-error-identification) on section508.gov, but didn't know if there was more available.
5. Is there guidance for formatting/styling content within modals?
6. Is there guidance for adding interactive components within a modal?

## Roadmap
1. Are there plans to add interactive pill/badge components?
2. Are there plans to add dropdown menus with multi-select and search bars that can be placed in-line in a table?

## Implementation
1. Is spacing in rems or pixels? How do we use rems if the site is specified in pixels? Assuming this may involve using design tokens in place of either.
2. What is the rationale for using a serif font in the headers of modals, when serif fonts are rarely used elsewhere?
3. How do you handle page zoom with modal content? How do you ensure that the controls of the modal remain visible at more zoom levels?

**Comment by mejiaj** [comment]
Hello, I'll do my best to provide some answers. I'll also ping the rest of the team to see if they can provide additional information.

The [website](https://designsystem.digital.gov/) is a great resource for these types of questions. If there's anything not covered here please let us know or create an [issue](https://github.com/uswds/uswds-site/issues/new/choose) so we can improve our docs!

## Guidance

1. Do you have an example for this use case? The Table component has guidance on [_When to consider something else_](https://designsystem.digital.gov/components/table/#when-to-consider-something-else) that might help. If not, we can create an issue to add it.
2. The [alert component](https://designsystem.digital.gov/components/alert/#when-to-use-the-alert-component) seems to fit this criteria.
3. I don't have anything specific, except match the visual order. I bet @alex-hull or @amycole501 have more insights.
4. We don't have any specific styling information related to content. Could you provide more detail?
5. Could you provide more detail?

## Roadmap

1. There's currently no roadmap item or item scheduled for pill/badge. There is this open issue with voting - [Change the component name "tag" to the more neutral "badge" · Issue #4612 · uswds/uswds](https://github.com/uswds/uswds/issues/4612).
2. Not at the moment, but feel free to create a [feature request](https://github.com/uswds/uswds/issues/new?assignees=&labels=Type%3A+Feature+Request&projects=&template=feature_request.yaml&title=USWDS+-+Feature%3A+%5BYOUR+TITLE%5D) with additional information. Once opened, we'll add the `Voting open` label and users can upvote (via 👍 ― _we'll use that to prioritize for next Roadmap planning_). Providing UX research or realistic use cases would help in validating this request. 

## Implementation

1. Tokens are a great way to keep your UI cohesive; see [Design Token guidance](https://designsystem.digital.gov/design-tokens/) & [Spacing unit guidance](https://designsystem.digital.gov/design-tokens/spacing-units/) for more information. Using `rem` values is the preferred method. There's a function called [`px-to-rem()`](https://github.com/uswds/uswds/blob/develop/packages/uswds-core/src/styles/functions/units/px-to-rem.scss) that lets you easily convert, but I recommend using tokens (where possible).
2. This was a design choice, I'm not aware of the specifics. The [Typography guidance](https://designsystem.digital.gov/components/typography/) says: 
      >"Government websites need clear and consistent headings, highly legible body paragraphs, clear labels, and easy-to-use input fields. Our default typefaces are designed for legibility and can adapt to a variety of visual tones." You can always update in the [global typography settings](https://designsystem.digital.gov/documentation/settings/#typography-settings).
3. This highlights the importance of using relative units like `rem` over `px` values.

**Reply by ShallieDragon** [reply]
Hello,

I am on the same team as the original poster. I can provide some context on a few of these.

### Guidance
1. An example from our work: we are looking to display columns of data fields about a record in our system. This is typically formatted as one column showing the label for the data field, and the other column showing the value for that field (for example, "Name: Jon Smith", "Role: Software Developer"). However, the use of a full-on table felt heavy-handed, as it would use a lot of extraneous lines and column headers to display what is essentially just read-only text. What would be the best way to display this kind of data using USWDS?
2. As I understand it, the Alert component does not provide guidance for more than static alerts. It doesn't mention fade-in/fade-out time, display duration, any interactive elements on the alert, placement on the page, allowance for placement over top of other elements, presence/absence of drop shadows, etc. Is there any advice on how to handle these aspects using USWDS?
3. (no comments, thank you for the help)
4. This can also be lumped in with 5. We are trying to allow users to modify the contents of a table, without requiring them to navigate away from that table. Specifically, allowing users to select one or more items from a list of choices, and having that choice reflected in the table. I had previously considered some kind of in-line control for editing things (e.g. dropdown menu), but none of them supported multi-select. So I considered a modal. However, the default modal component doesn't provide much guidance for interactivity and styling. It appears to only be used for simple confirmations. So, any suggestions for how we might allow users to perform these actions on a table would be appreciated. 

Let me know if you have any thoughts.

**Reply by amycole501** [reply]
Thanks for the follow up info, this was helpful. Here are my two cents on the following questions/situations:
2. Fading, display duration, and interactivity along with placement are all features that would warrant a customized accessibility audit on the page where the component was implemented. The alert component in its current state is useful for showing error messages in forms, for example, which need not be interactive or fade in or out. Is what you're suggesting something that would act more like a dynamic  feature the user would interact with or be able to dismiss?
4/5. Anytime you introduce interactivity with a visitor ensure it follows proper focus order and other WCAG standards. This [quick reference](https://www.w3.org/WAI/WCAG21/quickref/?currentsidebar=%23col_customize#top) can help ensure that whatever variations are implemented are accessible. Ensure you also test any type of interaction on a phone if your users are on mobile since tables are notorious for being difficult to read on mobile.

**Reply by mahoneycm** [reply]
1. It may depend on implementation, but one option could be to use [grid](https://designsystem.digital.gov/utilities/layout-grid/) to create two columns. 

For example:
![image](https://github.com/uswds/uswds/assets/25211650/252c4fc3-92df-49ef-9d66-953802488ee7)

```html
<div class="padding-2 grid-container">
  <div class="grid-row">
    <h2>Our team</h2>
  </div>
  <div class="grid-row">
    <div class="grid-col">
      <p class="text-bold">Name: Jon Smith</p>
    </div>
    <div class="grid-col">
      <p class="text-light">Role: Software Developer</p>
    </div>
  </div>
  <div class="grid-row">
    <div class="grid-col">
      <p class="text-bold">Name: Jane Doe</p>
    </div>
    <div class="grid-col">
      <p class="text-light">Role: User Experience</p>
    </div>
  </div>
</div>
```

If you want to maintain the data's relationship with each other, I imagine something like the [borderless table](https://designsystem.digital.gov/components/table/#borderless-table-2) might be better for you.

4/5. Our guidance generally recommends keeping modals simple and considering other options before resorting to modals. You can however use other interactive elements within modals, such as checkboxes, dropdowns / select elements if deemed necessary. I would just recommend doing ample UX consideration before implementing.

![image](https://github.com/uswds/uswds/assets/25211650/66472e38-59fb-4df6-8407-8861f9d56a0d)

Select elements also support the [multiple](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select#multiple) attribute
> This Boolean attribute indicates that multiple options can be selected in the list. If it is not specified, then only one option can be selected at a time. When multiple is specified, most browsers will show a scrolling list box instead of a single-line dropdown.

**Reply by ShallieDragon** [reply]
@amycole501 Regarding 2: yes, we were thinking of something more elaborate than alerts in a form. Like, site-level alerts, to inform users about the success or failure of actions. Currently, we have an implementation that displays a success alert when the user completes a particular gesture, as feedback that their action worked. We have it set to fade in and fade out automatically, and display for a certain duration. Further, it appears over the content, with a drop shadow, so that users can see the alert regardless of their scroll position. 

The main concern I have is that our implementation seems to be a significant departure from the original version on USWDS. We've had to take some creative liberties, and I was concerned that it might be negatively impacting accessibility or usability. I am relatively new to designing with WCAG standards in mind (so I appreciate the resources you've provided), so I've been leaning heavily on USWDS to ensure that a lot of our components meet accessibility guidelines by default.

**Reply by amycole501** [reply]
That's a valid concern; that type of customized situation would require a manual audit (and if possible, user experience test) for compliance once it's to the draft stage. We have been doing some work on creating critical checklists for teams to revalidate the components on their own systems to check for accessibility regression. We hope to have those posted to the site in the near future. The myriad of design variables are important to recheck, so hang onto that [W3C quick reference](https://www.w3.org/WAI/WCAG21/quickref/?currentsidebar=%23col_customize#top) as it'll walk you through the success criteria you need to test for.

**Reply by governmentSponsored** [reply]
@mejiaj regarding the design choice for serif vs. sans-serif in the model. Is that something that could be revisited and changed? 

Folks who use it on our end have some strong feelings about having serifs anywhere in the UI, as they feel it makes it more difficult to read and digest. They say this is especially true as it differs from the rest of the text in the UI that doesn't utilize serifs.

I know we can overwrite, and set fonts on our end, but we'd like to avoid that where possible.

**Reply by mejiaj** [reply]
@governmentSponsored that should be fairly easy to override. If it's not, we can create an issue to fix it.

**Comment by alex-hull** [comment]
Hello!

In reference to guidance question #3: The Section508 resource is a great starting point for understanding the general idea of tab order! While we do not have guidance on tab order per se, WCAG does have a few success criteria that could be of use: [Meaningful Sequence](https://www.w3.org/TR/WCAG20-TECHS/G57.html), [Focus Order](https://www.w3.org/TR/WCAG20-TECHS/G59.html#G59-tests) for interactive elements, and [Headings and Labels](https://www.w3.org/WAI/WCAG21/Understanding/headings-and-labels.html).  Each provides a user with assistive technology the ability to navigate a site, without it jumping between content in a non-sequential way. 

Another resource that may be able to provide a bit more insight could be the [screen reader survey](https://webaim.org/projects/screenreadersurvey9/#finding) performed by WebAIM this past year. While each screen reader user has their own methods of navigating sites, this provides a general overlook of the typical ways many assistive technology users are accessing websites.

**Comment by jaclinec** [comment]
Hi there!

For the guidance question 1, the closest thing I think USWDS offers that could be useful are [borderless tables](https://designsystem.digital.gov/components/table/#:~:text=Component%20code-,Borderless%20table,-A%20borderless%20table) which can be used when you are going for more simplicity and less lines. I'd also check out our guidance for [typography](https://designsystem.digital.gov/components/typography/) which has guidance for text alignment, line length, etc. If these resources aren't quite meeting your needs, let us know and we can create an issue to address what you're looking for.

**Comment by governmentSponsored** [comment]
Thank you @jaclinec @alex-hull @amycole501 @mahoneycm and @mejiaj for your thoughtful and insightful responses! They've been very helpful :)
